samples/api/swapchain_recreation/swapchain_recreation.cpp

```diff
@@ -0,0 +1,930 @@
+
+	// Only rebuild the swapchain if the dimensions have changed
+	if (surface_properties.currentExtent.width == swapchain_extents.width &&
+	    surface_properties.currentExtent.height == swapchain_extents.height &&
+	    desired_present_mode == current_present_mode)
+	{
+		return false;
+	}
+
+	init_swapchain();
+	return true;
+}
+
+void SwapchainRecreation::setup_frame()
+{
+	// For the frame we need:
+	// - A fence for the submission
+	// - A semaphore for image acquire
+	// - A semaphore for image present
+
+	// But first, pace the CPU.  Wait for frame N-2 to finish before starting recording of frame N.
+	submit_history_index = (submit_history_index + 1) % submit_history.size();
+	PerFrame &frame      = submit_history[submit_history_index];
+	if (frame.submit_fence != VK_NULL_HANDLE)
+	{
+		vkWaitForFences(get_device_handle(), 1, &frame.submit_fence, true, UINT64_MAX);
+
+		// Reset/recycle resources, they are no longer in use.
+		recycle_fence(frame.submit_fence);
+		recycle_semaphore(frame.acquire_semaphore);
+		vkResetCommandPool(get_device_handle(), frame.primary_command_pool, 0);
+
+		// Destroy any garbage that's associated with this submission.
+		for (SwapchainObjects &garbage : frame.swapchain_garbage)
+		{
+			cleanup_swapchain_objects(garbage);
+		}
+		frame.swapchain_garbage.clear();
+
+		// Note that while the submission fence, the semaphore it waited on and the command
+		// pool its command was allocated from are guaranteed to have finished execution,
+		// there is no guarantee that the present semaphore is not in use.
+		//
+		// This is because the fence wait above ensures that the submission _before_ present
+		// is finished, but makes no guarantees as to the state of the present operation
+		// that follows.  The present semaphore is queued for garbage collection when
+		// possible after present, and is not kept as part of the submit history.
+		assert(frame.present_semaphore == VK_NULL_HANDLE);
+	}
+
+	frame.submit_fence      = get_fence();
+	frame.acquire_semaphore = get_semaphore();
+	frame.present_semaphore = get_semaphore();
+
+	if (frame.primary_command_pool == VK_NULL_HANDLE)
+	{
+		VkCommandPoolCreateInfo cmd_pool_info{VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO};
+		cmd_pool_info.flags            = VK_COMMAND_POOL_CREATE_TRANSIENT_BIT;
+		cmd_pool_info.queueFamilyIndex = queue->get_family_index();
+		VK_CHECK(vkCreateCommandPool(get_device_handle(), &cmd_pool_info, nullptr, &frame.primary_command_pool));
+	}
+}
+
+void SwapchainRecreation::render(uint32_t index)
+{
+	PerFrame &frame = submit_history[submit_history_index];
+
+	VkCommandBuffer             cmd;
+	VkCommandBufferAllocateInfo cmd_buf_info{VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO};
+	cmd_buf_info.commandPool        = frame.primary_command_pool;
+	cmd_buf_info.level              = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+	cmd_buf_info.commandBufferCount = 1;
+	VK_CHECK(vkAllocateCommandBuffers(get_device_handle(), &cmd_buf_info, &cmd));
```

Comment:
I believe the cause of the slowing down over time is because you are allocating a new command buffer every frame but never freeing it.  You reset the command pool in setup_frame() but that just puts any command buffers in the pool back into 'initial' state, it doesn't free them. The list of command buffers in the pool just keeps growing, which causes vkResetCommandPool to take longer and longer for me.

Reply:
Oops, fixed. Let me know if it's ok now

Reply:
> Oops, fixed. Let me know if it's ok now

It no longer slows over time, thanks.